drgn_tools/dentry.py

```diff
@@ -381,24 +381,44 @@ def __file_type(mode: Object) -> str:
     return "UNKN"
 
 
-def ls(prog: Program, directory: str, count: bool = False) -> None:
+def ls(
+    prog: Program,
+    directory: str,
+    parent_dentry: Object,
+    depth: int,
+    count: bool = False,
+) -> None:
     """
     Print dentry children, like the ls command
     :param directory: directory to print children of
     :param count: when true, only print counts (not the full contents)
     """
-    dentries = dentry_for_each_child(path_lookup(prog, directory).dentry)
+    if depth < 0:
+        return
+
+    subdirs = []
+    dentries = dentry_for_each_child(
+        parent_dentry if parent_dentry else path_lookup(prog, directory).dentry
+    )
 
+    print(f"{directory}:")
     pos = neg = 0
     for i, dentry in enumerate(dentries):
+        dentry_type = "positive"
         path = dentry_path_any_mount(dentry).decode()
         if dentry_is_negative(dentry):
             neg += 1
+            dentry_type = "negative"
         else:
+            if stat.S_ISDIR(int(dentry.d_inode.i_mode)):
+                subdirs.append([path, dentry])
             pos += 1
         if not count:
-            print(f"{i:05d} {path}")
-    print(f"{pos} positive, {neg} negative dentries")
+            print(f"\t{i:05d} {path} ({dentry_type})")
+    print(f"\t{pos} positive, {neg} negative dentries")
```

Comment:
This isn't part of your change, but I think printing the full path is a bit too much for each dentry. For deeply nested filesystem paths that's a waste. Let's just print the dentry name!

So instead of using `dentry_path_any_mount()` we can do:

```
name = dentry.d_name.name.string().decode()
...
print(f" .... {name} ...")
```

And then, for the subdirectories, we can do:

```
subdirs.append((f"{directory}/{name}", dentry))
```